packages/common/src/sendTransaction.ts

```diff
@@ -0,0 +1,88 @@
+import {
+  Account,
+  CallParameters,
+  Chain,
+  Client,
+  SendTransactionParameters,
+  Transport,
+  WriteContractReturnType,
+} from "viem";
+import { call, sendTransaction as viem_sendTransaction } from "viem/actions";
+import pRetry from "p-retry";
+import { debug as parentDebug } from "./debug";
+import { getNonceManager } from "./getNonceManager";
+import { parseAccount } from "viem/accounts";
+
+const debug = parentDebug.extend("sendTransaction");
+
+// TODO: migrate away from this approach once we can hook into viem's nonce management: https://github.com/wagmi-dev/viem/discussions/1230
+
+export async function sendTransaction<
+  TChain extends Chain | undefined,
+  TAccount extends Account | undefined,
+  TChainOverride extends Chain | undefined
+>(
+  client: Client<Transport, TChain, TAccount>,
+  request: SendTransactionParameters<TChain, TAccount, TChainOverride>
+): Promise<WriteContractReturnType> {
+  const account_ = request.account ?? client.account;
+  if (!account_) {
+    // TODO: replace with viem AccountNotFoundError once its exported
+    throw new Error("No account provided");
+  }
+  const account = parseAccount(account_);
+
+  const nonceManager = await getNonceManager({
+    client,
+    address: account.address,
+  });
+
+  async function prepare(): Promise<SendTransactionParameters<TChain, TAccount, TChainOverride>> {
+    if (request.gas) {
+      debug("gas provided, skipping simulate", request.to);
+      return request;
+    }
+
+    debug("simulating tx to", request.to);
+    await call(client, {
+      ...request,
+      blockTag: "pending",
+      account,
+    } as CallParameters<TChain>);
+
+    // TODO: estimate gas
```

Comment:
ahh so I thought viem internally did both a call/simulate + estimate gas, but looking closer at [`sendTransaction`](https://github.com/wagmi-dev/viem/blob/main/src/actions/wallet/sendTransaction.ts) and [`prepareTransactionRequest`](https://github.com/wagmi-dev/viem/blob/main/src/actions/wallet/prepareTransactionRequest.ts), it looks like it's just a gas estimate (likely because we don't need the simulated return value in this context)

will see about updating, but may need to also have to add gas retry handling if there's too much time between calling this function and the tx getting sent to the mem pool (due to queue and/or nonce retries)

Reply:
for context: I was initially just doing the simulate, not the gas estimate, because I wanted to know if it was safe to increment the nonce but still allow for the gas to be computed just before sending the tx

but since gas estimate is ~the same cost as simulate and we can attempt to send the tx with potentially outdated gas and save a call, this seems worth it (assuming we have proper retry handling for bad gas)

Reply:
looks like viem does use the return data of the simulate call for better error messages in case of failures/reverts

Reply:
I attempted to do this, along with consolidating behavior, but ran into issues with being able to parse errors from viem. I've already burned a few hours on this, so I'd like to pause and come back to this another time in a follow-up: https://github.com/latticexyz/mud/pull/1717/commits/97fd6d792ad5a87aad47de30a54bd946730def61

Reply:
sounds good. Do you think it's a viem bug or something on our side? If the former, is there a repro we could share with them?

Reply:
I wasn't able to confirm if it was a viem issue or not. Using their StackBlitz starter kit didn't seem to surface it. But it definitely happens within the MUD repo, even though everything is pinned to exactly the same version (tried clearing all node modules and reinstalling and also tried bumping to a different exactly version, with no change in behavior I was seeing).